pallets/settlement/src/lib.rs

```diff
@@ -1609,101 +1644,124 @@ impl<T: Config> Module<T> {
-        }
+        // Chanrges the weight consumed so far
+        weight_meter.check_accrue(<T as Config>::WeightInfo::base_execute_instruction(
+            instruction_legs.len() as u32,
+        ));
 
-        match frame_storage_with_transaction(|| {
-            Self::release_asset_locks_and_transfer_pending_legs(instruction_id, &instruction_legs)
+        // Verifies if the venue is allowed for all tickers in the instruction
+        Self::ensure_allowed_venue(&instruction_legs, venue_id, weight_meter)?;
+
+        // Attempts to release the locks and transfer all fungible an non fungible assets
+        if let Err(leg_id) = frame_storage_with_transaction(|| {
+            Self::release_asset_locks_and_transfer_pending_legs(
+                instruction_id,
+                &instruction_legs,
+                weight_meter,
+            )
         })? {
-            Ok(_) => {
-                Self::deposit_event(RawEvent::InstructionExecuted(
-                    SettlementDID.as_id(),
-                    instruction_id,
-                ));
-            }
-            Err(leg_id) => {
-                Self::deposit_event(RawEvent::LegFailedExecution(
-                    SettlementDID.as_id(),
-                    instruction_id,
-                    leg_id,
-                ));
-                Self::deposit_event(RawEvent::InstructionFailed(
-                    SettlementDID.as_id(),
-                    instruction_id,
-                ));
-                // Unclaim receipts for the failed transaction so that they can be reused
-                Self::unsafe_unclaim_receipts(instruction_id, &instruction_legs);
-                return Err(Error::<T>::InstructionFailed.into());
-            }
+            Self::deposit_event(RawEvent::LegFailedExecution(
+                SettlementDID.as_id(),
+                instruction_id,
+                leg_id,
+            ));
+            Self::deposit_event(RawEvent::InstructionFailed(
+                SettlementDID.as_id(),
+                instruction_id,
+            ));
+            // Unclaim receipts for the failed transaction so that they can be reused
+            Self::unsafe_unclaim_receipts(instruction_id, &instruction_legs);
+            return Err(Error::<T>::InstructionFailed.into());
         }
 
-        Ok(instruction_legs.len().try_into().unwrap_or_default())
+        Self::deposit_event(RawEvent::InstructionExecuted(
+            SettlementDID.as_id(),
+            instruction_id,
+        ));
+        Ok(())
     }
 
     fn release_asset_locks_and_transfer_pending_legs(
         instruction_id: InstructionId,
         instruction_legs: &[(LegId, LegV2)],
+        weight_meter: &mut WeightMeter,
     ) -> TransactionOutcome<Result<Result<(), LegId>, DispatchError>> {
         Self::unchecked_release_locks(instruction_id, instruction_legs);
+        // Charges the weight consumed
+        let transfer_data = Self::get_transfer_data(&instruction_legs).unwrap();
```

Comment:
Please don't use panicking `unwrap` here.  Only benchmarks and tests are allowed to panic.

Reply:
`get_transfer_data` doesn't have to return a result. Since this is only called after the instruction legs are already in the storage, all checks have already been made. The condition inside the function was duplicating one of the checks. I fixed it now.